release/src/router/softcenter_ext/perp/perp/man/perpok.8
.\" perpok.8
.\" wcm, 2009.11.10 - 2013.01.09
.\" ===
.TH perpok 8 "January 2013" "perp-2.07" "persistent process supervision"
.SH NAME
perpok \- check a
.BR perpd (8)
service
.SH SYNOPSIS
.B perpok [\-hV] [\-b
.I basedir
.B ] [\-u
.I secs
.B ] [\-v]
.I sv
.SH DESCRIPTION
.B perpok
is used to check that the service
.I sv
is actively running under supervision of 
.BR perpd (8).
.PP
The argument
.I sv
should be a service definition directory supervised by
.BR perpd (8).
.B perpok
will look for
.I sv
in the directory given with the
.B \-b
option,
or as set in the
.B PERP_BASE
environmental variable,
or in the current directory if neither of the previous is given.
.PP
.B perpok
returns 0 to indicate success if the definition directory
.I sv
exists and is under the active supervision of
.BR perpd (8).
.PP
The
.B \-u
option may be given to specify a minimum uptime in
.IR secs .
In this case,
.B perpok
will further check that the main service itself is running,
is not resetting,
does not want down,
and that its uptime has been at least
.I secs
seconds.
If all these conditions are met,
.B perpok
returns 0 to indicate success.
.PP
If the conditions for success are not met,
.B perpok
exits non-zero.
.PP
.B perpok
is intended primarily as a utility for boolean testing in scripting environments.
See the
.BR perpstat (8)
and
.BR perpls (8)
utilities for more detailed status reporting.
.SH OPTIONS
.TP
.B \-b basedir
Base directory.
Sets the base directory containing the service definition
.IR sv .
If not set,
.B perpok
will use the value set in the variable PERP_BASE,
or the current directory if neither of these are defined.
On a normal system,
the base directory will be
.IR /etc/perp .
.TP
.B \-h
Help.
Print a brief usage message to stderr and exit.
.TP
.B \-u secs
Uptime.
Normally
.B perpok
checks only that the
.I sv
directory is active (exists and is sticky),
and that
.BR perpd (8)
is actively supervising it.
This option extends the checks performed by
.B perpok
to test that the main service itself is running,
is not resetting,
does not want down,
and that its uptime has been at least
.I secs
seconds.
Normally it is sensible to set
.I secs
to some small value, such as 2 or 3,
which is
just long enough to check that the service is not cycling on restarts.
.TP
.B \-V
Version.
Print the version number to stderr and exit.
.TP
.B \-v
Verbose.
Normally
.B perpok
exits silently,
as it is designed primarily for runscripts that might use it to perform simple dependency checks
prior to starting their own services.
This option causes
.B perpok
to print a message reporting either success or failure to stderr on exit.
.SH EXAMPLES
.B perpok
may be used as a basic dependency checking utility within service runscripts:
.PP
.RS
.nf
#!/bin/sh
if test ${1} = "start" ; then
  if ! perpok -u3 foo ; then
    echo "dependency foo not running"
    exit 1
  fi
  # otherwise foo ok, continue to start service:
  exec /usr/bin/bar ...
fi
.fi
.RE
.PP
This example shows a runscript for a ``bar'' service that uses
.B perpok
to require a dependency on another service named ``foo''.
If foo has not been running for at least 3 seconds,
the runscript exits immediately.
The
.BR perpd (8)
supervisor running this service will then continue to try restarting it at one second intervals,
until the
.B perpok
dependency check succeeds.
.PP
Note that, when using
.B perpok
in a normal collection of services running under
.BR perpd (8),
PERP_BASE will always be properly predefined in the runscript environment.
.PP
.SH EXIT STATUS
.B perpok
exits with the following values:
.TP
0
Success.
The service
.I sv
is active and a
.BR perpd (8)
supervisor has been found running for it.
Additionally,
if the
.B \-u
option has been given,
the main service itself has been found running,
not resetting,
not wanting down,
and with an uptime of at least
.I secs
seconds.
.TP
1
Failure.
The service directory
.I sv
was found not to be active,
or a
.BR perpd (8)
supervisor was not found running for it,
or the additional constraints imposed for uptime checking
with the
.B \-u
option were not met.
.TP
100
Usage error.
For unknown options, missing arguments, or if
.I sv
does not name a directory.
Prints a brief diagnostic message to stderr on exit.
.TP
111
System error.
Unexpected failures during system calls, privilege and/or resource problems,
or configuration errors in the
.BR perpd (8)
system.
Prints a brief diagnostic message to stderr on exit.
.SH CAVEATS
For use as a dependency tool,
.B perpok
by itself is of course limited.
As illustrated in the example section above,
.B perpok
may find a dependency running okay at startup,
but there is no guarantee that it will continue to run for as long
as the dependent service requires it.
.PP
A collection of services running under
.BR perpd (8)
tends to be reliable,
so that the use of
.B perpok
for basic dependency checking at startup is normally adequate.
Interrelated services may also use
.B perpok
in combination with runscript
.B reset
targets to implement a system of stricter dependency controls. 
.SH AUTHOR
Wayne Marshall, http://b0llix.net/perp/
.SH SEE ALSO
.nh
.BR perp_intro (8),
.BR perpboot (8),
.BR perpctl (8),
.BR perpd (8),
.BR perpetrate (5),
.BR perphup (8),
.BR perpls (8),
.BR perpstat (8),
.BR sissylog (8),
.BR tinylog (8)
.\" EOF perpok.8


release/src/router/perp/perp/perpok.c
/* perpok.c
** perp: persistent process supervision
** perp 2.0: single process scanner/supervisor/controller
** perpok: query and return service "ok"
** (ipc client query to perpd server)
** wcm, 2009.11.09 - 2011.01.27
** ===
*/

/* libc: */
#include <stdlib.h>

/* unix: */
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/socket.h>
#include <sys/stat.h>

/* lasanga: */
#include "buf.h"
#include "cstr.h"
#include "domsock.h"
#include "fd.h"
#include "nextopt.h"
#include "nfmt.h"
#include "nuscan.h"
#include "pidlock.h"
#include "pkt.h"
#include "sysstr.h"
#include "tain.h"
#include "upak.h"
#include "uchar.h"

#include "perp_common.h"
#include "perp_stderr.h"


static const char *progname = NULL;
static const char prog_usage[] = "[-hV] [-b basedir] [-u secs] [-v] sv";


#define  report_fail(...) \
  {\
      if(verbose){\
          eputs(progname, ": ", svdir, ": fail: ", __VA_ARGS__);\
      }\
      die(1);\
  }

#define  report_ok(...) \
  {\
      if(verbose){\
          eputs(progname, ": ", svdir, ": ok: ",  __VA_ARGS__);\
      }\
      die(0);\
  }


int
main(int argc, char *argv[])
{
  nextopt_t       nopt = nextopt_INIT(argc, argv, ":hVb:u:v");
  char            opt;
  int             verbose = 0;
  const char     *basedir = NULL;
  uint32_t        arg_upsecs = 0;
  const char     *z;
  const char     *svdir = NULL;
  struct stat     st;
  char            pathbuf[256];
  size_t          n;
  int             fd_conn;
  pkt_t           pkt = pkt_INIT(2, 'Q', 16);
  uchar_t        *status = pkt_data(pkt);
  pid_t           pid_main;
  tain_t          when_main;
  uint64_t        uptime_main;
  uchar_t         flags;
  tain_t          now;

  progname = nextopt_progname(&nopt);
  while((opt = nextopt(&nopt))){
      char optc[2] = {nopt.opt_got, '\0'};
      switch(opt){
      case 'h': usage(); die(0); break;
      case 'V': version(); die(0); break;
      case 'b': basedir = nopt.opt_arg; break;
      case 'u':
          z = nuscan_uint32(&arg_upsecs, nopt.opt_arg);
          if(*z != '\0'){
              fatal_usage("non-numeric argument found for option -", optc, " ", nopt.opt_arg);
          }
          break;
      case 'v': ++verbose; break;
      case ':':
          fatal_usage("missing argument for option -", optc);
          break;
      case '?':
          if(nopt.opt_got != '?'){
              fatal_usage("invalid option -", optc);
          }
          /* else fallthrough: */
      default : die_usage(); break;
      }
  }

  argc -= nopt.arg_ndx;
  argv += nopt.arg_ndx;

  svdir = argv[0];
  if(svdir == NULL){
      fatal_usage("missing argument");
  }

  if(!basedir)
      basedir = getenv("PERP_BASE");
  if(!basedir)
      basedir = ".";


  if(chdir(basedir) != 0){
      fatal_syserr("failure chdir() to ", basedir);
  }

  if(stat(svdir, &st) == -1){
      fatal_syserr("failure stat() on service directory ", svdir);
  }

  if(! S_ISDIR(st.st_mode)){
      fatal_usage("argument not a directory: ", svdir);
  }

  if(!(S_ISVTX & st.st_mode)){
      report_fail("service directory ", svdir, " not activated");
  }

  /* connect to control socket: */
  n = cstr_vlen(basedir, "/", PERP_CONTROL, "/", PERPD_SOCKET);
  if(!(n < sizeof pathbuf)){
      errno = ENAMETOOLONG;
      fatal_syserr("failure locating perpd control socket ",
                   basedir, "/", PERP_CONTROL, "/", PERPD_SOCKET);
  }
  cstr_vcopy(pathbuf, basedir, "/", PERP_CONTROL, "/", PERPD_SOCKET);
  fd_conn = domsock_connect(pathbuf);
  if(fd_conn == -1){
      if(errno == ECONNREFUSED){
          fatal_syserr("perpd not running on control socket ", pathbuf);
      } else {
          fatal_syserr("failure connecting to perpd control socket ", pathbuf);
      }
  }

  /* uptime compared to now: */
  tain_now(&now);

  /* status query packet: */ 
  upak_pack(pkt_data(pkt), "LL", (uint64_t)st.st_dev, (uint64_t)st.st_ino);

  if(pkt_write(fd_conn, pkt, 0) == -1){
      fatal_syserr("failure pkt_write() to perpd control socket");
  }

  if(pkt_read(fd_conn, pkt, 0) == -1){
      fatal_syserr("failure pkt_read() from perpd control socket");
  }

  /* done with connection: */
  close(fd_conn);

  if(pkt[0] != 2){
      fatal(111, "unknown protocol found in reply from perpd control socket");
  }
  if(pkt[1] != 'S'){
      if(pkt[1] == 'E'){
          errno = (int)upak32_unpack(&pkt[3]);
          fatal_syserr("error reported in reply from perpd control socket");
      } else {
          fatal(111, "unknown packet reply type from perpd control socket");
      }
  }

  if(arg_upsecs == 0){
  /* basic test complete: */
      report_ok("service is activated/ok");
  }

  /* else continue extended testing... */

  pid_main = upak32_unpack(&status[30]);
  if(!(pid_main > 0)){
      report_fail("service not running (pid is 0)");
  }

  flags = status[46];
  if(flags & SUBSV_FLAG_ISRESET){
      report_fail("service is resetting");
  }
  if(flags & SUBSV_FLAG_WANTDOWN){
      report_fail("service wants down");
  }

  tain_unpack(&when_main, &status[34]);
  uptime_main = tain_uptime(&now, &when_main);
  if(uptime_main < (uint64_t)arg_upsecs){
      report_fail("service uptime not met");
  }

  /* okie dokie! */
  report_ok("service uptime ok");

  /* not reached: */
  die(0);
}


/* eof: perpok.c */
